### DISCUSSION TITLE
mpi4py ctypes wrapper for Fortran 2008

### DISCUSSION THREAD

**Discussion by apoorv-2-8** [mpi4py/mpi4py#276] [Q&A] [answered]
Hi,
I was writing a "Hello World" example using ctypes wraper for Fortran using MPI_F08 module in fortran file. I was able to successfully run it on i5 Laptop using ubuntu WSL2. But same files give error using ubuntu WSL2 on XEON processor machine. The  error code it throws is 139. I am just learning MPI and experimenting, any guidance would be very helpful.
Regards
Apoorv

The two file (python and fortran ) are
![image](https://user-images.githubusercontent.com/81359062/199775947-d67046c8-dc47-481d-a5f2-f8fc2cde09ac.png)
![image](https://user-images.githubusercontent.com/81359062/199776150-b0d3d042-df75-4224-b0f7-0d41d8623dad.png)

The output on I5 laptop is
![image](https://user-images.githubusercontent.com/81359062/199776383-c81de671-2a2d-4f69-bf97-7ed52d32ec7f.png)

**Comment by dalcinl** [comment, marked as answer]
Next time, please do not post screenshots. Just paste the test and use proper markdown formatting. Otherwise, you make things harder to try and play around for those trying to help you.

The code you are using would be appropriate for C toYou should learn elsewehre make it work in any platform with any MPI implementation. But it is not correct for Fortran 77 or Fortran 90/08.

First, you should uncomment `bind(C)` for maximum portability.

Next, you need to do things the following way in Python :

```python
MPI_Fint = ctypes.c_int
MPI_Fint_p = ctypes.POINTER(MPI_Fint)
lib.sayhello.restype = None
lib.sayhello.argtypes = [MPI_Fint_p]  # by default, Fortran pass arguments by address

comm = MPI.COMM_WORLD
fcomm = MPI_Fint(comm.py2f())
lib.sayhello(fcomm)
```

This is not the only way to do things. In your fortran code, you could use:
```fortran
type(MPI_Comm), intent(in), value :: comm
```
Note the use of `value` to pass the argument by value rather than address.

As you pass the comm argument by value, then in Python you just need:
```
MPI_Fint = ctypes.c_int
lib.sayhello.restype = None
lib.sayhello.argtypes = [MPI_Fint]  # now the argument is passed by value, not by address
```

**Reply by apoorv-2-8** [reply]
@dalcinl 
Thank you for your reply, it helped to resolve the issue.

**Comment by dalcinl** [comment]
@apoorv-2-8 I've put together a couple examples with the stuff we discussed here. The new examples are in this (branch)[https://github.com/mpi4py/mpi4py/tree/feature/demo-ctypes-fortran], commit 29c898efe7c9b5ceaf6919ff1f0bbe079f0f3178. Could you please take a quick look and tell me whether this is good enough to merge into master?

**Reply by apoorv-2-8** [reply]
@dalcinl , Thank you for adding the Fortran 2008 ctypes example. I worked well for me on Ubuntu wsl. I think its good to merge into master

**Reply by dalcinl** [reply]
Great! Thanks for the feedback. Commit ed084db1b304a29088402695e85a24c96ee63232 pushed  to master.
